server/src/main/java/org/elasticsearch/index/mapper/DotExpandingXContentParser.java

```diff
@@ -73,14 +74,15 @@ private void expandDots() throws IOException {
             if (subpaths.length == 1 && field.endsWith(".") == false) {
                 return;
             }
-            XContentLocation location = delegate().getTokenLocation();
-            Token token = delegate().nextToken();
-            if (token == Token.START_OBJECT || token == Token.START_ARRAY) {
-                parsers.push(new DotExpandingXContentParser(new XContentSubParser(delegate()), delegate(), subpaths, location));
-            } else if (token == Token.END_OBJECT || token == Token.END_ARRAY) {
+            Token token = delegate.nextToken();
+            if (token == Token.END_OBJECT || token == Token.END_ARRAY) {
                 throw new IllegalStateException("Expecting START_OBJECT or START_ARRAY or VALUE but got [" + token + "]");
             } else {
-                parsers.push(new DotExpandingXContentParser(new SingletonValueXContentParser(delegate()), delegate(), subpaths, location));
+                var location = delegate.getTokenLocation();
+                var parser = (token == Token.START_OBJECT || token == Token.START_ARRAY)
+                    ? new XContentSubParser(delegate)
+                    : new SingletonValueXContentParser(delegate);
+                parsers.push(new DotExpandingXContentParser(parser, delegate, subpaths, location));
```

Comment:
This caused the test failure at: `org.elasticsearch.index.mapper.DotExpandingXContentParserTests#testGetTokenLocation`